### DISCUSSION TITLE
posts/python-coroutine-programming-1/

### DISCUSSION THREAD

**Discussion by giscus** [humminglab/blog.humminglab.io#2] [Announcements]
# posts/python-coroutine-programming-1/

Python2 와 비교하여 python3의 가장 돋보이는 killer feature 는 비동기 프로그래밍 지원이라고 할 수 있다. 이를 위하여 python 3.4에 asyncio 모듈이 추가되었고, python 3.5 에는 native coroutine 지원을 위한 async, await 키워드가 추가되었다. 이들 기능을 이용하면 javascript나 다른 언어에서 지원하는 비동기 프로그래밍의 장점을 python 에서도 사용할 수 있다. 즉, 이벤트 방식이지만 blocking 방식의 프로그래밍 처럼 sequential 하게 코드를 작성할 수 있어, 단일 thread로 수만개의 네트워크 연결을 처리하는 서버를 오류 가능성을 최소화 하면서, 보다 편하게 개발할 수 있다.

https://blog.humminglab.io/posts/python-coroutine-programming-1/

**Comment by cakel** [comment]
좋은 글 감사합니다. 차근차근 따라하다가 Iterator 의 두번째 예제 코드에 오류가 나서 막혔는데 코드를 약간 수정하니 해결되었습니다.
```
>>> next(c)
Traceback (most recent call last):
  File "<string>", line 19, in <module>
TypeError: 'Counter' object is not an iterator
```
```
c = Counter()
```
에서 다음 라인에
```
c = iter(c)
```
를 추가 하니 정상 출력되었습니다. 파이썬 버전(v3.10.5) 에 따라 동작 방법이 다른지는 모르겠지만, Iterable 객체에서 Iterator 로 변환하는 과정이 있어야 했나봅니다.

문제를 이해하는데, 추가적으로 https://bluese05.tistory.com/55 페이지를 참고 했습니다.

---

다음 예제에서 바로 막히네요.
```
for i in c:
TypeError: 'Iterator' object is not iterable
```

Iterator 클래스에 아래와 같이 임시방편으로 __iter__ 메소드를 추가해서 Iterable 하게 만들어보니, 정상 출력이 되었습니다.
```
    def __iter__(self):
        return self
```

제가 파이썬 초보라서 잘못 작성된걸 수도 있습니다. 작성하신 예제를 제 PC에서 동작되게 수정하면 아래와 같습니다.
```
class Counter(object):
    def __iter__(self):
        iter = Iterator()
        return iter

class Iterator(object):
    def __init__(self):
        self.index = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.index > 10:
            raise StopIteration
        n = self.index * 2
        self.index += 1
        return n

c = Counter()
c = iter(c)

for i in c:
    print(i)
```

결과
```
0
2
4
6
8
10
12
14
16
18
20
```

**Reply by aqwerf** [reply]
안녕하세요. 잘못된 부분을 자세하게 알려주셔서 감사합니다. 
Iterator object를 얻을 수 있는 `iter()`를 호출해 주는 것이 맞습니다. 

```python
c = iter (c)
```

https://bluese05.tistory.com/55 에서 언급한 것과 같이 iterable, iterator 로 구분을 한다면 다음처럼 됩니다.

- `__iter__()` method가 있는 class는 iterable
- `__next__()` method가 있는 class는 iterator

```python
class Counter(object):
    def __iter__(self):
        iter = Iterator()
        return iter

class Iterator(object):
    def __init__(self):
        self.index = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.index > 10:
            raise StopIteration
        n = self.index * 2
        self.index += 1
        return n

c = Counter()
i = iter(c)

print(next(i))
```

위에서 보면 `c`는 iterable object 입니다. 이를 builtin `iter()` 함수로 얻게 되는 것이고요.
`iter()`함수의 역할이 parameter로 전달받은 object의 `__iter__()` method 를 호출하여 얻는 과정입니다.

i는 iterator object 입니다. 이를 builtin `next()` 함수로 얻게 되는 것이고요. 
`next()` 함수의 역할이 paramter로 전달받은 object의 `__next__()` method 를 호출해서 얻는 과정입니다. 

```python
c = Counter()
for n in c:
    print(n)
```

for...in 루프는 iterable을 취급하므로 위처럼 iterable object (`__iter__`를 가진) 를 넣어주면 됩니다. 

본문도 수정하였습니다.
